Change to CHI misses some updates required for windows

<!-- NOTE: If your issue is a security concern, please send an email to security@gitea.io instead of opening a public issue -->

<!--
    1. Please speak English, this is the language all maintainers can speak and write.
    2. Please ask questions or configuration/deploy problems on our Discord
       server (https://discord.gg/gitea) or forum (https://discourse.gitea.io).
    3. Please take a moment to check that your issue doesn't already exist.
    4. Make sure it's not mentioned in the FAQ (https://docs.gitea.io/en-us/faq)
    5. Please give all relevant information below for bug reports, because
       incomplete details will be handled as an invalid report.
-->

- Gitea version (or commit ref): latest
- Git version:
- Operating system: Windows
  <!-- Please include information on whether you built gitea yourself, used one of our downloads or are using some other package -->
  <!-- Please also tell us how you are running gitea, e.g. if it is being run from docker, a command-line, systemd etc. --->
  <!-- If you are using a package or systemd tell us what distribution you are using -->
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [X ] No
- Log gist:
<!-- It really is important to provide pertinent logs -->
<!-- Please read https://docs.gitea.io/en-us/logging-configuration/#debugging-problems -->
<!-- In addition, if your problem relates to git commands set `RUN_MODE=dev` at the top of app.ini -->

## Description

Got this error when I run go run main.go

```C:\Users\arnna\Desktop\CodeShare\App>go run main.go
# code.gitea.io/gitea/modules/auth/sso
modules\auth\sso\sspi_windows.go:69:27: undefined: ctx
modules\auth\sso\sspi_windows.go:79:56: undefined: ctx
modules\auth\sso\sspi_windows.go:82:37: undefined: ctx
modules\auth\sso\sspi_windows.go:87:15: undefined: ctx
modules\auth\sso\sspi_windows.go:88:3: undefined: ctx
modules\auth\sso\sspi_windows.go:89:3: undefined: ctx
modules\auth\sso\sspi_windows.go:90:3: undefined: ctx
modules\auth\sso\sspi_windows.go:94:37: undefined: ctx
modules\auth\sso\sspi_windows.go:113:24: too many arguments in call to s.newUser
modules\auth\sso\sspi_windows.go:113:25: undefined: ctx
modules\auth\sso\sspi_windows.go:113:24: too many errors```
## Screenshots

<!-- **If this issue involves the Web Interface, please include a screenshot** -->

--- a/.drone.yml
+++ b/.drone.yml
@@ -28,11 +28,22 @@ steps:
     image: golang:1.15
     commands:
       - make lint-backend
-      - GOOS=windows make golangci-lint vet
     environment:
       GOPROXY: https://goproxy.cn # proxy.golang.org is blocked in China, this proxy is not
       GOSUMDB: sum.golang.org
       TAGS: bindata sqlite sqlite_unlock_notify
+
+  - name: lint-backend-windows
+    pull: always
+    image: golang:1.15
+    commands:
+      - make golangci-lint vet
+    environment:
+      GOPROXY: https://goproxy.cn # proxy.golang.org is blocked in China, this proxy is not
+      GOSUMDB: sum.golang.org
+      TAGS: bindata sqlite sqlite_unlock_notify
+      GOOS: windows
+      GOARCH: amd64
 
   - name: lint-backend-gogit
     pull: always

--- a/.drone.yml
+++ b/.drone.yml
@@ -28,6 +28,7 @@ steps:
     image: golang:1.15
     commands:
       - make lint-backend
+      - GOOS=windows make golangci-lint vet
     environment:
       GOPROXY: https://goproxy.cn # proxy.golang.org is blocked in China, this proxy is not
       GOSUMDB: sum.golang.org

--- a/modules/auth/sso/basic.go
+++ b/modules/auth/sso/basic.go
@@ -47,7 +47,7 @@ func (b *Basic) IsEnabled() bool {
 // "Authorization" header of the request and returns the corresponding user object for that
 // name/token on successful validation.
 // Returns nil if header is empty or validation fails.
-func (b *Basic) VerifyAuthData(req *http.Request, store DataStore, sess SessionStore) *models.User {
+func (b *Basic) VerifyAuthData(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
 	baHead := req.Header.Get("Authorization")
 	if len(baHead) == 0 {
 		return nil

--- a/modules/auth/sso/interface.go
+++ b/modules/auth/sso/interface.go
@@ -40,5 +40,5 @@ type SingleSignOn interface {
 	// or a new user object (with id = 0) populated with the information that was found
 	// in the authentication data (username or email).
 	// Returns nil if verification fails.
-	VerifyAuthData(http *http.Request, store DataStore, sess SessionStore) *models.User
+	VerifyAuthData(http *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User
 }

--- a/modules/auth/sso/oauth2.go
+++ b/modules/auth/sso/oauth2.go
@@ -114,7 +114,7 @@ func (o *OAuth2) IsEnabled() bool {
 // or the "Authorization" header and returns the corresponding user object for that ID.
 // If verification is successful returns an existing user object.
 // Returns nil if verification fails.
-func (o *OAuth2) VerifyAuthData(req *http.Request, store DataStore, sess SessionStore) *models.User {
+func (o *OAuth2) VerifyAuthData(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
 	if !models.HasEngine {
 		return nil
 	}

--- a/modules/auth/sso/reverseproxy.go
+++ b/modules/auth/sso/reverseproxy.go
@@ -60,7 +60,7 @@ func (r *ReverseProxy) IsEnabled() bool {
 // If a username is available in the "setting.ReverseProxyAuthUser" header an existing
 // user object is returned (populated with username or email found in header).
 // Returns nil if header is empty.
-func (r *ReverseProxy) VerifyAuthData(req *http.Request, store DataStore, sess SessionStore) *models.User {
+func (r *ReverseProxy) VerifyAuthData(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
 	username := r.getUserName(req)
 	if len(username) == 0 {
 		return nil

--- a/modules/auth/sso/session.go
+++ b/modules/auth/sso/session.go
@@ -39,7 +39,7 @@ func (s *Session) IsEnabled() bool {
 // VerifyAuthData checks if there is a user uid stored in the session and returns the user
 // object for that uid.
 // Returns nil if there is no user uid stored in the session.
-func (s *Session) VerifyAuthData(req *http.Request, store DataStore, sess SessionStore) *models.User {
+func (s *Session) VerifyAuthData(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
 	user := SessionUser(sess)
 	if user != nil {
 		return user

--- a/modules/auth/sso/sspi_windows.go
+++ b/modules/auth/sso/sspi_windows.go
@@ -5,10 +5,11 @@
 package sso
 
 import (
-	"code.gitea.io/gitea/modules/templates"
 	"errors"
 	"net/http"
 	"strings"
+
+	"code.gitea.io/gitea/modules/templates"
 
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/base"
@@ -82,7 +83,9 @@ func (s *SSPI) VerifyAuthData(req *http.Request, w http.ResponseWriter, store Da
 		// Include the user login page in the 401 response to allow the user
 		// to login with another authentication method if SSPI authentication
 		// fails
-		//addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
+		store.GetData()["Flash"] = map[string]string{
+			"ErrMsg": err.Error(),
+		}
 
 		store.GetData()["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
 		store.GetData()["EnableSSPI"] = true
@@ -228,24 +231,6 @@ func sanitizeUsername(username string, cfg *models.SSPIConfig) string {
 	username = replaceSeparators(username, cfg)
 	return username
 }
-
-/*
-// TODO flash err not implemented for chi
-// addFlashErr adds an error message to the Flash object mapped to a macaron.Context
-func addFlashErr(ctx *macaron.Context, err string) {
-	fv := ctx.GetVal(reflect.TypeOf(&session.Flash{}))
-	if !fv.IsValid() {
-		return
-	}
-	flash, ok := fv.Interface().(*session.Flash)
-	if !ok {
-		return
-	}
-	flash.Error(err)
-	ctx.Data["Flash"] = flash
-}
-
-*/
 
 // init registers the SSPI auth method as the last method in the list.
 // The SSPI plugin is expected to be executed last, as it returns 401 status code if negotiation

--- a/modules/auth/sso/sspi_windows.go
+++ b/modules/auth/sso/sspi_windows.go
@@ -40,6 +40,7 @@ var (
 // On successful authentication returns a valid user object.
 // Returns nil if authentication fails.
 type SSPI struct {
+	rnd *render.Render
 }
 
 // Init creates a new global websspi.Authenticator object
@@ -47,7 +48,18 @@ func (s *SSPI) Init() error {
 	config := websspi.NewConfig()
 	var err error
 	sspiAuth, err = websspi.New(config)
-	return err
+	if err != nil {
+		return err
+	}
+	s.rnd = render.New(render.Options{
+		Extensions:    []string{".tmpl"},
+		Directory:     "templates",
+		Funcs:         templates.NewFuncMap(),
+		Asset:         templates.GetAsset,
+		AssetNames:    templates.GetAssetNames,
+		IsDevelopment: setting.RunMode != "prod",
+	})
+	return nil
 }
 
 // Free releases resources used by the global websspi.Authenticator object
@@ -86,19 +98,10 @@ func (s *SSPI) VerifyAuthData(req *http.Request, w http.ResponseWriter, store Da
 		store.GetData()["Flash"] = map[string]string{
 			"ErrMsg": err.Error(),
 		}
-
 		store.GetData()["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
 		store.GetData()["EnableSSPI"] = true
 
-		rnd := render.New(render.Options{
-			Extensions:    []string{".tmpl"},
-			Directory:     "templates",
-			Funcs:         templates.NewFuncMap(),
-			Asset:         templates.GetAsset,
-			AssetNames:    templates.GetAssetNames,
-			IsDevelopment: setting.RunMode != "prod",
-		})
-		err := rnd.HTML(w, 401, string(tplSignIn), templates.BaseVars().Merge(store.GetData()))
+		err := s.rnd.HTML(w, 401, string(tplSignIn), templates.BaseVars().Merge(store.GetData()))
 		if err != nil {
 			log.Error("%v", err)
 		}

--- a/modules/auth/sso/sspi_windows.go
+++ b/modules/auth/sso/sspi_windows.go
@@ -9,12 +9,11 @@ import (
 	"net/http"
 	"strings"
 
-	"code.gitea.io/gitea/modules/templates"
-
 	"code.gitea.io/gitea/models"
 	"code.gitea.io/gitea/modules/base"
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/templates"
 
 	gouuid "github.com/google/uuid"
 	"github.com/quasoft/websspi"

--- a/modules/auth/sso/sspi_windows.go
+++ b/modules/auth/sso/sspi_windows.go
@@ -5,9 +5,9 @@
 package sso
 
 import (
+	"code.gitea.io/gitea/modules/templates"
 	"errors"
 	"net/http"
-	"reflect"
 	"strings"
 
 	"code.gitea.io/gitea/models"
@@ -15,11 +15,9 @@ import (
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
 
-	"gitea.com/macaron/macaron"
-	"gitea.com/macaron/session"
-
 	gouuid "github.com/google/uuid"
 	"github.com/quasoft/websspi"
+	"github.com/unrolled/render"
 )
 
 const (
@@ -65,8 +63,8 @@ func (s *SSPI) IsEnabled() bool {
 // If authentication is successful, returs the corresponding user object.
 // If negotiation should continue or authentication fails, immediately returns a 401 HTTP
 // response code, as required by the SPNEGO protocol.
-func (s *SSPI) VerifyAuthData(req *http.Request, store DataStore, sess SessionStore) *models.User {
-	if !s.shouldAuthenticate(ctx) {
+func (s *SSPI) VerifyAuthData(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) *models.User {
+	if !s.shouldAuthenticate(req) {
 		return nil
 	}
 
@@ -76,22 +74,36 @@ func (s *SSPI) VerifyAuthData(req *http.Request, store DataStore, sess SessionSt
 		return nil
 	}
 
-	userInfo, outToken, err := sspiAuth.Authenticate(req, ctx.Resp)
+	userInfo, outToken, err := sspiAuth.Authenticate(req, w)
 	if err != nil {
 		log.Warn("Authentication failed with error: %v\n", err)
-		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+		sspiAuth.AppendAuthenticateHeader(w, outToken)
 
 		// Include the user login page in the 401 response to allow the user
 		// to login with another authentication method if SSPI authentication
 		// fails
-		addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
-		ctx.Data["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
-		ctx.Data["EnableSSPI"] = true
-		ctx.HTML(401, string(tplSignIn))
+		//addFlashErr(ctx, ctx.Tr("auth.sspi_auth_failed"))
+
+		store.GetData()["EnableOpenIDSignIn"] = setting.Service.EnableOpenIDSignIn
+		store.GetData()["EnableSSPI"] = true
+
+		rnd := render.New(render.Options{
+			Extensions:    []string{".tmpl"},
+			Directory:     "templates",
+			Funcs:         templates.NewFuncMap(),
+			Asset:         templates.GetAsset,
+			AssetNames:    templates.GetAssetNames,
+			IsDevelopment: setting.RunMode != "prod",
+		})
+		err := rnd.HTML(w, 401, string(tplSignIn), templates.BaseVars().Merge(store.GetData()))
+		if err != nil {
+			log.Error("%v", err)
+		}
+
 		return nil
 	}
 	if outToken != "" {
-		sspiAuth.AppendAuthenticateHeader(ctx.Resp, outToken)
+		sspiAuth.AppendAuthenticateHeader(w, outToken)
 	}
 
 	username := sanitizeUsername(userInfo.Username, cfg)
@@ -110,7 +122,7 @@ func (s *SSPI) VerifyAuthData(req *http.Request, store DataStore, sess SessionSt
 			log.Error("User '%s' not found", username)
 			return nil
 		}
-		user, err = s.newUser(ctx, username, cfg)
+		user, err = s.newUser(username, cfg)
 		if err != nil {
 			log.Error("CreateUser: %v", err)
 			return nil
@@ -118,8 +130,8 @@ func (s *SSPI) VerifyAuthData(req *http.Request, store DataStore, sess SessionSt
 	}
 
 	// Make sure requests to API paths and PWA resources do not create a new session
-	if !isAPIPath(ctx) && !isAttachmentDownload(ctx) {
-		handleSignIn(ctx, sess, user)
+	if !isAPIPath(req) && !isAttachmentDownload(req) {
+		handleSignIn(w, req, sess, user)
 	}
 
 	return user
@@ -146,7 +158,7 @@ func (s *SSPI) shouldAuthenticate(req *http.Request) (shouldAuth bool) {
 	if path == "/user/login" {
 		if req.FormValue("user_name") != "" && req.FormValue("password") != "" {
 			shouldAuth = false
-		} else if ctx.Req.FormValue("auth_with_sspi") == "1" {
+		} else if req.FormValue("auth_with_sspi") == "1" {
 			shouldAuth = true
 		}
 	} else if isInternalPath(req) {
@@ -217,6 +229,8 @@ func sanitizeUsername(username string, cfg *models.SSPIConfig) string {
 	return username
 }
 
+/*
+// TODO flash err not implemented for chi
 // addFlashErr adds an error message to the Flash object mapped to a macaron.Context
 func addFlashErr(ctx *macaron.Context, err string) {
 	fv := ctx.GetVal(reflect.TypeOf(&session.Flash{}))
@@ -230,6 +244,8 @@ func addFlashErr(ctx *macaron.Context, err string) {
 	flash.Error(err)
 	ctx.Data["Flash"] = flash
 }
+
+*/
 
 // init registers the SSPI auth method as the last method in the list.
 // The SSPI plugin is expected to be executed last, as it returns 401 status code if negotiation

--- a/modules/auth/sso/user.go
+++ b/modules/auth/sso/user.go
@@ -12,7 +12,7 @@ import (
 
 // SignedInUser returns the user object of signed user.
 // It returns a bool value to indicate whether user uses basic auth or not.
-func SignedInUser(req *http.Request, ds DataStore, sess SessionStore) (*models.User, bool) {
+func SignedInUser(req *http.Request, w http.ResponseWriter, ds DataStore, sess SessionStore) (*models.User, bool) {
 	if !models.HasEngine {
 		return nil, false
 	}
@@ -22,7 +22,7 @@ func SignedInUser(req *http.Request, ds DataStore, sess SessionStore) (*models.U
 		if !ssoMethod.IsEnabled() {
 			continue
 		}
-		user := ssoMethod.VerifyAuthData(req, ds, sess)
+		user := ssoMethod.VerifyAuthData(req, w, ds, sess)
 		if user != nil {
 			_, isBasic := ssoMethod.(*Basic)
 			return user, isBasic

--- a/modules/context/context.go
+++ b/modules/context/context.go
@@ -309,7 +309,7 @@ func Contexter() macaron.Handler {
 		}
 
 		// Get user from session if logged in.
-		ctx.User, ctx.IsBasicAuth = sso.SignedInUser(ctx.Req.Request, ctx, ctx.Session)
+		ctx.User, ctx.IsBasicAuth = sso.SignedInUser(ctx.Req.Request, c.Resp, ctx, ctx.Session)
 
 		if ctx.User != nil {
 			ctx.IsSigned = true

--- a/modules/graceful/manager_windows.go
+++ b/modules/graceful/manager_windows.go
@@ -73,7 +73,7 @@ func (g *Manager) start() {
 
 	// Make SVC process
 	run := svc.Run
-	isInteractive, err := svc.IsAnInteractiveSession()
+	isInteractive, err := svc.IsWindowsService()
 	if err != nil {
 		log.Error("Unable to ascertain if running as an Interactive Session: %v", err)
 		return
@@ -81,7 +81,9 @@ func (g *Manager) start() {
 	if isInteractive {
 		run = debug.Run
 	}
-	go run(WindowsServiceName, g)
+	go func() {
+		_ = run(WindowsServiceName, g)
+	}()
 }
 
 // Execute makes Manager implement svc.Handler

--- a/modules/log/console_windows.go
+++ b/modules/log/console_windows.go
@@ -23,7 +23,7 @@ func enableVTMode(console windows.Handle) bool {
 	// https://docs.microsoft.com/en-us/windows/console/setconsolemode
 	// It only works on windows 10. Earlier terminals will fail with an err which we will
 	// handle to say don't color
-	mode = mode | windows.ENABLE_VIRTUAL_TERMINAL_PROCESSING
+	mode |= windows.ENABLE_VIRTUAL_TERMINAL_PROCESSING
 	err = windows.SetConsoleMode(console, mode)
 	return err == nil
 }

--- a/routers/routes/recovery.go
+++ b/routers/routes/recovery.go
@@ -75,7 +75,7 @@ func Recovery() func(next http.Handler) http.Handler {
 					}
 
 					// Get user from session if logged in.
-					user, _ := sso.SignedInUser(req, &store, sess)
+					user, _ := sso.SignedInUser(req, w, &store, sess)
 					if user != nil {
 						store.Data["IsSigned"] = true
 						store.Data["SignedUser"] = user

